### PR TITLE
Wallet CLI Implementation

### LINKED ISSUE #537: Wallet CLI commands
**Task summary**

dependent on #465 and just provides a CLI interface for the functionality exposed over rpc

commands options reference (from lotus):
     new          Generate a new key of the given type
     list         List wallet address
     balance      Get account balance
     export       export keys
     import       import keys
     default      Get default wallet address
     set-default  Set default wallet address

**Specification reference**

**Other information and links**

### PR DESCRIPTION
**Pull request by connormullett**
**Summary of changes**
Changes introduced in this pull request:
- Wallet commands mirrored from lotus

**Reference issue to close (if applicable)**

Closes 
#537 
#1118 

**Other information and links**

### CODE REVIEW

**Comment by connormullett** [comment]
Update: 7/11 commands working. Import, export, sign, verify are still WIP

**Comment by connormullett** [comment]
Marking RFR. The 2 checks that failed are unrelated to the changes introduced in this PR

File: forest/src/cli/wallet_cmd.rs
Comment (creativcoder) [inline forest/src/cli/wallet_cmd.rs:194]: `eprintln!()` and let's print the value maybe

File: forest/src/cli/wallet_cmd.rs
Comment (creativcoder) [inline forest/src/cli/wallet_cmd.rs:29]: nit: "The address of the account to check"

File: forest/src/cli/wallet_cmd.rs
Comment (creativcoder) [inline forest/src/cli/wallet_cmd.rs:51]: nit: "default"

File: forest/src/cli/wallet_cmd.rs
Comment (connormullett) [reply forest/src/cli/wallet_cmd.rs:29]: Good catch haha

File: forest/src/cli/wallet_cmd.rs
Comment (connormullett) [reply forest/src/cli/wallet_cmd.rs:51]: Another good catch lol

File: forest/src/cli/wallet_cmd.rs
Comment (connormullett) [reply forest/src/cli/wallet_cmd.rs]: Made to use `cli_error_and_die`

File: key_management/src/keystore.rs
Comment (cryptoquick) [inline key_management/src/keystore.rs:66]: It's hard to tell is this trait still needed?

File: node/rpc-client/src/client.rs
Comment (cryptoquick) [inline node/rpc-client/src/client.rs:190]: I recommend a `*` import here. We use all of them from this module.

File: node/rpc-client/src/client.rs
Comment (cryptoquick) [inline node/rpc-client/src/client.rs:242]: This needs to be a Params and Result

File: node/rpc-client/src/client.rs
Comment (cryptoquick) [inline node/rpc-client/src/client.rs:222]: Try to make Result types match up with those in the `rpc_api` in addition to Params. Each one needs to follow the pattern.

File: node/rpc-client/src/lib.rs
Comment (cryptoquick) [inline node/rpc-client/src/lib.rs:7]: Is this `pub mod` necessary when we're making all members in it public below?

File: node/rpc-client/src/wallet_ops.rs
Comment (cryptoquick) [inline node/rpc-client/src/wallet_ops.rs:12]: Make sure all Params and Results types are used in this file.

File: node/rpc/src/lib.rs
Comment (cryptoquick) [inline node/rpc/src/lib.rs:16]: Is this ever used anywhere else? Does it need to be public?

File: vm/address/src/lib.rs
Comment (cryptoquick) [inline vm/address/src/lib.rs:335]: Do we still need this trait?

File: vm/message/src/signed_message.rs
Comment (cryptoquick) [inline vm/message/src/signed_message.rs:135]: Do we still need this trait?

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Great work on documentation. We need more of that. However, I noticed you're not using the `rpc_api` fully. Let me know if you need help understanding how to best make use of it, however, the patterns should be complete enough in the other files that implement it in `rpc_client`.

**Comment by cryptoquick** [comment]
I noticed one additional issue after testing this.

Screenshot from 2021-06-10 16-13-23

I would recommend making secp256k1 the default, like lotus.

Otherwise, I tested out the new, list, and default commands, and they all worked great.

I also double-checked that read permissions return a 403, while write permissions work as expected, and they all do, which is also good! The permissions list I added is working.

File: node/rpc-client/src/client.rs
Comment (connormullett) [reply node/rpc-client/src/client.rs:190]: :+1:

File: key_management/src/keystore.rs
Comment (connormullett) [reply key_management/src/keystore.rs:66]: Not anymore, good catch. Was used for debugging

File: node/rpc-client/src/client.rs
Comment (connormullett) [reply node/rpc-client/src/client.rs:222]: Yup, skipped over these. Think they were from before they existed? Idk. Fixed them up

File: vm/message/src/signed_message.rs
Comment (connormullett) [reply vm/message/src/signed_message.rs:135]: Nope. All of these Debug traits were for checking output before making them pretty. Completely forgot to remove

File: node/rpc/src/lib.rs
Comment (connormullett) [reply node/rpc/src/lib.rs:16]: The other command modules used the raw function name and imported directly from `rpc_client`. I wanted to explicitly have `wallet_ops::..` in the function call for clarity as to where the function lives. Removed to keep the pattern with other commands

File: node/rpc-client/src/client.rs
Comment (cryptoquick) [inline node/rpc-client/src/client.rs:216]: Missed one: https://github.com/ChainSafe/forest/blob/e4e6711b62925a90b813272a33ebc7eb99be2862/node/rpc-api/src/lib.rs#L299

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Just one more thing you missed, after this, I think we're good. I've resolved the conversations that either have been addressed or will be done through additional `rpc_api` refactoring.

**Review by cryptoquick — APPROVED** [review]
Just tested each command. Checked means it works:
- [x] wallet default (without a wallet configured)
- [x] wallet new
- [x] wallet new secp256k1
- [x] wallet new bls
- [x] wallet has (tested wtih bls, secp256k1, and garbage addresses)
- [x] wallet list
- [x] wallet default (defaults to first wallet made)
- [x] wallet set-default (tested with bls, secp256k1, and garbage addresses)
  - garbage address causes unhandled panic in forest CLI (will make a separate issue)
  - `FULLNODE_API_INFO="eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.j5yGdcMgSAQd6raXLsryx-3I3fvL05QYh6BSKoNZtBA:/ip4/127.0.0.1/tcp/1234/http" forest wallet set-default f3va7krfg4svarx4uqfudab2z3bzbnjn6kipuahc5t3bp5jboowdnx2lj6vysbyaalkibeu7zhnala4sbku3gaasdf
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidLength', forest/src/cli/wallet_cmd.rs:160:63`
  - See: #1166 
- [x] wallet export (bls and secp256k1)
- [x] wallet import (handles existing key)
- [x] wallet sign -m 'f00f' -a <address> (and verified the bytes differen than if signing with `-m 'f33f'`)
- [x] wallet verify - Fails with error decoding address. Will make a separate issue so it won't necessarily block merging this PR, since it doesn't break existing functionality. See: #1165

File: docs/CLI.md
Comment (ec2) [inline docs/CLI.md:25]:
```suggestion
This will use IPv4, TCP, and HTTP when communicating with the RPC API. The admin token can be found when starting
```

File: docs/CLI.md
Comment (ec2) [inline docs/CLI.md:26]:
```suggestion
the Forest daemon. This will be needed to create tokens with certain permissions such as read, write, sign, or admin.
```

File: docs/CLI.md
Comment (ec2) [inline docs/CLI.md:33]:
```suggestion
Retrieve the FIL balance of a given address
```

**Review by ec2 — CHANGES_REQUESTED** [review]
Some minor typos and things. As well as what @cryptoquick said. Not sure what happened... It was working last time we paired on it.... Though I think we should probably fix that before merging.

**Comment by cryptoquick** [comment]
@ec2 I'm helping Connor get this across the finish line so it doesn't languish too long while he's got his current obligations to Uncle Sam. I've made changes based on your feedback.

I'd like @connormullett to take on the two fixes in #1165 and #1166, but there's a lot of value in this PR already, and it doesn't break anything that existed before, I think? I'll leave it up to him to either fix it, or merge and fix via the tracking issue.

**Comment by ec2** [comment]
> @ec2 I'm helping Connor get this across the finish line so it doesn't languish too long while he's got his current obligations to Uncle Sam. I've made changes based on your feedback.
> 
> 
> 
> I'd like @connormullett to take on the two fixes in #1165 and #1166, but there's a lot of value in this PR already, and it doesn't break anything that existed before, I think? I'll leave it up to him to either fix it, or merge and fix via the tracking issue.

Makes sense. Forgot about uncle sam.

**Review by ec2 — APPROVED** [review]

**Comment by cryptoquick** [comment]
I merged it for Connor. I told him to take all the time he needs to address the remaining issues. It'll be good for me to build the net API on top of this, anyway.

**Pull request merged**
